frame/uniques/src/lib.rs

```diff
@@ -0,0 +1,1289 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Unique (Assets) Module
+//!
+//! A simple, secure module for dealing with non-fungible assets.
+//!
+//! ## Related Modules
+//!
+//! * [`System`](../frame_system/index.html)
+//! * [`Support`](../frame_support/index.html)
+
+// Ensure we're `no_std` when compiling for Wasm.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub mod weights;
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+#[cfg(test)]
+pub mod mock;
+#[cfg(test)]
+mod tests;
+
+mod types;
+pub use types::*;
+
+use sp_std::prelude::*;
+use sp_runtime::{RuntimeDebug, ArithmeticError, traits::{Zero, StaticLookup, Saturating}};
+use codec::{Encode, Decode, HasCompact};
+use frame_support::traits::{Currency, ReservableCurrency, BalanceStatus::Reserved};
+use frame_system::Config as SystemConfig;
+
+pub use weights::WeightInfo;
+pub use pallet::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+	use super::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T, I = ()>(_);
+
+	#[pallet::config]
+	/// The module configuration trait.
+	pub trait Config<I: 'static = ()>: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self, I>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Identifier for the class of asset.
+		type ClassId: Member + Parameter + Default + Copy + HasCompact;
+
+		/// The type used to identify a unique asset within an asset class.
+		type InstanceId: Member + Parameter + Default + Copy + HasCompact + From<u16>;
+
+		/// The currency mechanism, used for paying for reserves.
+		type Currency: ReservableCurrency<Self::AccountId>;
+
+		/// The origin which may forcibly create or destroy an asset or otherwise alter privileged
+		/// attributes.
+		type ForceOrigin: EnsureOrigin<Self::Origin>;
+
+		/// The basic amount of funds that must be reserved for an asset class.
+		type ClassDeposit: Get<DepositBalanceOf<Self, I>>;
+
+		/// The basic amount of funds that must be reserved for an asset instance.
+		type InstanceDeposit: Get<DepositBalanceOf<Self, I>>;
+
+		/// The basic amount of funds that must be reserved when adding metadata to your asset.
+		type MetadataDepositBase: Get<DepositBalanceOf<Self, I>>;
+
+		/// The basic amount of funds that must be reserved when adding an attribute to an asset.
+		type AttributeDepositBase: Get<DepositBalanceOf<Self, I>>;
+
+		/// The additional funds that must be reserved for the number of bytes store in metadata,
+		/// either "normal" metadata or attribute metadata.
+		type DepositPerByte: Get<DepositBalanceOf<Self, I>>;
+
+		/// The maximum length of a name or symbol stored on-chain.
```

Comment:
```suggestion
		/// The maximum length of data stored on-chain.
```
Also maybe doesn't makes sense to call it `StringLimit`, once the `data` field can store any binary format, not just a UTF-8 string.

Reply:
If this data MUST be a string, that string limit is ambiguous once string binary size != char count:
```rust
fn main() {
    let yes = "y̆es";
    println!("length: {}", yes.len());
    println!(" chars: {}", yes.chars().count());
}
```
output:
```
length: 4
 chars: 3
```

So the user may mistakenly think the `StringLimit` bounds the `data.chars().count()` when it actually limits the `data.len()`.